## Data Analysis and Visualization - Canada Immigration Data

### The Dataset: Immigration to Canada from 1980 to 2013 <a id="1"></a>

Dataset Source: [International migration flows to and from selected countries - The 2015 revision](https://www.un.org/development/desa/pd/data/international-migration-flows).

The dataset contains annual data on the flows of international immigrants as recorded by the countries of destination. The data presents both inflows and outflows according to the place of birth, citizenship or place of previous / next residence both for foreigners and nationals. The current study presents data pertaining to 45 countries.

 The Canada Immigration dataset can be fetched from <a href="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork/Data%20Files/Canada.xlsx">here</a>.

 Source : © IBM Corporation 2020. All rights reserved.

Import required modules.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt 

Read the dataset.

In [2]:
df = pd.read_csv("01 Canada Immigration Raw Data.csv",)
df.head()

Unnamed: 0        Type    Coverage          OdName  AREA AreaName   REG  \
0           0  Immigrants  Foreigners     Afghanistan   935     Asia  5501   
1           1  Immigrants  Foreigners         Albania   908   Europe   925   
2           2  Immigrants  Foreigners         Algeria   903   Africa   912   
3           3  Immigrants  Foreigners  American Samoa   909  Oceania   957   
4           4  Immigrants  Foreigners         Andorra   908   Europe   925   

           RegName  DEV             DevName  ...  2004  2005  2006  2007  \
0    Southern Asia  902  Developing regions  ...  2978  3436  3009  2652   
1  Southern Europe  901   Developed regions  ...  1450  1223   856   702   
2  Northern Africa  902  Developing regions  ...  3616  3626  4807  3623   
3        Polynesia  902  Developing regions  ...     0     0     1     0   
4  Southern Europe  901   Developed regions  ...     0     0     1     1   

   2008  2009  2010  2011  2012  2013  
0  2111  1746  1758  2203  2635  2004  
1   560   716   561   539   620   603  
2  4005  5393  4752  4325  3774  4331  
3     0     0     0     0     0     0  
4     0     0     0     0     1     1  

[5 rows x 44 columns]

In [3]:
df =df.drop("Unnamed: 0", axis =1)

In [4]:
df.head()

Type    Coverage          OdName  AREA AreaName   REG  \
0  Immigrants  Foreigners     Afghanistan   935     Asia  5501   
1  Immigrants  Foreigners         Albania   908   Europe   925   
2  Immigrants  Foreigners         Algeria   903   Africa   912   
3  Immigrants  Foreigners  American Samoa   909  Oceania   957   
4  Immigrants  Foreigners         Andorra   908   Europe   925   

           RegName  DEV             DevName  1980  ...  2004  2005  2006  \
0    Southern Asia  902  Developing regions    16  ...  2978  3436  3009   
1  Southern Europe  901   Developed regions     1  ...  1450  1223   856   
2  Northern Africa  902  Developing regions    80  ...  3616  3626  4807   
3        Polynesia  902  Developing regions     0  ...     0     0     1   
4  Southern Europe  901   Developed regions     0  ...     0     0     1   

   2007  2008  2009  2010  2011  2012  2013  
0  2652  2111  1746  1758  2203  2635  2004  
1   702   560   716   561   539   620   603  
2  3623  4005  5393  4752  4325  3774  4331  
3     0     0     0     0     0     0     0  
4     1     0     0     0     0     1     1  

[5 rows x 43 columns]

In [5]:
df.shape

(195, 43)

#### Excel File source

df_can = pd.read_excel(
    'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork/Data%20Files/Canada.xlsx',
    sheet_name='Canada by Citizenship',
    skiprows=range(20),
    skipfooter=2)
print('Data read into a pandas dataframe!')

In [6]:
df.info(verbose=False)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195 entries, 0 to 194
Columns: 43 entries, Type to 2013
dtypes: int64(37), object(6)
memory usage: 65.6+ KB


In [7]:
df.columns

Index(['Type', 'Coverage', 'OdName', 'AREA', 'AreaName', 'REG', 'RegName',
       'DEV', 'DevName', '1980', '1981', '1982', '1983', '1984', '1985',
       '1986', '1987', '1988', '1989', '1990', '1991', '1992', '1993', '1994',
       '1995', '1996', '1997', '1998', '1999', '2000', '2001', '2002', '2003',
       '2004', '2005', '2006', '2007', '2008', '2009', '2010', '2011', '2012',
       '2013'],
      dtype='object')

In [8]:
df.index

RangeIndex(start=0, stop=195, step=1)

Note: The default type of intance variables `index` and `columns` are **NOT** `list`.


In [9]:
print(type(df.columns))
print(type(df.index))

<class 'pandas.core.indexes.base.Index'>
<class 'pandas.core.indexes.range.RangeIndex'>


To get the index and columns as lists, we can use the `tolist()` method.


In [10]:
df.columns.to_list()

['Type',
 'Coverage',
 'OdName',
 'AREA',
 'AreaName',
 'REG',
 'RegName',
 'DEV',
 'DevName',
 '1980',
 '1981',
 '1982',
 '1983',
 '1984',
 '1985',
 '1986',
 '1987',
 '1988',
 '1989',
 '1990',
 '1991',
 '1992',
 '1993',
 '1994',
 '1995',
 '1996',
 '1997',
 '1998',
 '1999',
 '2000',
 '2001',
 '2002',
 '2003',
 '2004',
 '2005',
 '2006',
 '2007',
 '2008',
 '2009',
 '2010',
 '2011',
 '2012',
 '2013']

In [ ]:
df.index.to_list()

In [11]:
print(type(df.columns.tolist()))
print(type(df.index.tolist()))

<class 'list'>
<class 'list'>


Let's clean the data set to remove a few unnecessary columns. We can use *pandas* `drop()` method as follows:


In [12]:
# in pandas axis=0 represents rows (default) and axis=1 represents columns.
df.drop(['AREA', 'REG', 'DEV', 'Type', 'Coverage'], axis =1, inplace =True)
df.head(3)

OdName AreaName          RegName             DevName  1980  1981  \
0  Afghanistan     Asia    Southern Asia  Developing regions    16    39   
1      Albania   Europe  Southern Europe   Developed regions     1     0   
2      Algeria   Africa  Northern Africa  Developing regions    80    67   

   1982  1983  1984  1985  ...  2004  2005  2006  2007  2008  2009  2010  \
0    39    47    71   340  ...  2978  3436  3009  2652  2111  1746  1758   
1     0     0     0     0  ...  1450  1223   856   702   560   716   561   
2    71    69    63    44  ...  3616  3626  4807  3623  4005  5393  4752   

   2011  2012  2013  
0  2203  2635  2004  
1   539   620   603  
2  4325  3774  4331  

[3 rows x 38 columns]

In [14]:
df.rename(columns ={"OdName":"Country", "RegName": "Region", "AreaName": "Continent"}, inplace =True)
df.head(2)

Country Continent           Region             DevName  1980  1981  \
0  Afghanistan      Asia    Southern Asia  Developing regions    16    39   
1      Albania    Europe  Southern Europe   Developed regions     1     0   

   1982  1983  1984  1985  ...  2004  2005  2006  2007  2008  2009  2010  \
0    39    47    71   340  ...  2978  3436  3009  2652  2111  1746  1758   
1     0     0     0     0  ...  1450  1223   856   702   560   716   561   

   2011  2012  2013  
0  2203  2635  2004  
1   539   620   603  

[2 rows x 38 columns]

We will also add a 'Total' column that sums up the total immigrants by country over the entire period 1980 - 2013, as follows:


If your DataFrame contains `non-numeric columns` and you want to ensure they are excluded, the first approach (using `select_dtypes`) is usually the most straightforward.

In [15]:
df['Total'] = df.select_dtypes(include ='number').sum(axis=1)
df['Total']

0      58639
1      15699
2      69439
3          6
4         15
       ...  
190    97146
191        2
192     2985
193     1677
194     8598
Name: Total, Length: 195, dtype: int64

Now let's check for the missing values.

In [17]:
df.isnull().value_counts()

Country  Continent  Region  DevName  1980   1981   1982   1983   1984   1985   1986   1987   1988   1989   1990   1991   1992   1993   1994   1995   1996   1997   1998   1999   2000   2001   2002   2003   2004   2005   2006   2007   2008   2009   2010   2011   2012   2013   Total
False    False      False   False    False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False    195
Name: count, dtype: int64

In [18]:
df.describe()

1980          1981          1982          1983          1984  \
count    195.000000    195.000000    195.000000    195.000000    195.000000   
mean     508.394872    566.989744    534.723077    387.435897    376.497436   
std     1949.588546   2152.643752   1866.997511   1204.333597   1198.246371   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%       13.000000     10.000000     11.000000     12.000000     13.000000   
75%      251.500000    295.500000    275.000000    173.000000    181.000000   
max    22045.000000  24796.000000  20620.000000  10015.000000  10170.000000   

              1985         1986          1987          1988          1989  \
count   195.000000   195.000000    195.000000    195.000000    195.000000   
mean    358.861538   441.271795    691.133333    714.389744    843.241026   
std    1079.309600  1225.576630   2109.205607   2443.606788   2555.048874   
min       0.000000     0.000000      0.000000      0.000000      0.000000   
25%       0.000000     0.500000      0.500000      1.000000      1.000000   
50%      17.000000    18.000000     26.000000     34.000000     44.000000   
75%     197.000000   254.000000    434.000000    409.000000    508.500000   
max    9564.000000  9470.000000  21337.000000  27359.000000  23795.000000   

       ...          2005          2006          2007          2008  \
count  ...    195.000000    195.000000    195.000000    195.000000   
mean   ...   1320.292308   1266.958974   1191.820513   1246.394872   
std    ...   4425.957828   3926.717747   3443.542409   3694.573544   
min    ...      0.000000      0.000000      0.000000      0.000000   
25%    ...     28.500000     25.000000     31.000000     31.000000   
50%    ...    210.000000    218.000000    198.000000    205.000000   
75%    ...    832.000000    842.000000    899.000000    934.500000   
max    ...  42584.000000  33848.000000  28742.000000  30037.000000   

               2009          2010          2011          2012          2013  \
count    195.000000    195.000000    195.000000    195.000000    195.000000   
mean    1275.733333   1420.287179   1262.533333   1313.958974   1320.702564   
std     3829.630424   4462.946328   4030.084313   4247.555161   4237.951988   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%       36.000000     40.500000     37.500000     42.500000     45.000000   
50%      214.000000    211.000000    179.000000    233.000000    213.000000   
75%      888.000000    932.000000    772.000000    783.000000    796.000000   
max    29622.000000  38617.000000  36765.000000  34315.000000  34129.000000   

               Total  
count     195.000000  
mean    32867.451282  
std     91785.498686  
min         1.000000  
25%       952.000000  
50%      5018.000000  
75%     22239.500000  
max    691904.000000  

[8 rows x 35 columns]

### Slicing the dataframe

Let's try filtering on the list of countries ('Country') and the data for years: 1980 - 1985.


In [19]:
df[['Country', '1980', '1981', '1982', '1983', '1984', '1985']] # returns a dataframe

Country  1980  1981  1982  1983  1984  1985
0       Afghanistan    16    39    39    47    71   340
1           Albania     1     0     0     0     0     0
2           Algeria    80    67    71    69    63    44
3    American Samoa     0     1     0     0     0     0
4           Andorra     0     0     0     0     0     0
..              ...   ...   ...   ...   ...   ...   ...
190        Viet Nam  1191  1829  2162  3404  7583  5907
191  Western Sahara     0     0     0     0     0     0
192           Yemen     1     2     1     6     0    18
193          Zambia    11    17    11     7    16     9
194        Zimbabwe    72   114   102    44    32    29

[195 rows x 7 columns]

Notice that 'Country' is string, and the years are integers.For the sake of consistency, we will convert all column names to string later on.

There are main 2 ways to select rows:

```python
    df.loc[label]    # filters by the labels of the index/column
    df.iloc[index]   # filters by the positions of the index/column

Before we proceed, notice that the default index of the dataset is a numeric range from 0 to 194. This makes it very difficult to do a query by a specific country. For example to search for data on Japan, we need to know the corresponding index value.

This can be fixed very easily by setting the 'Country' column as the index using `set_index()` method.

In [20]:
df.set_index('Country', inplace=True)
# The opposite of set is reset. So to reset the index, we can use df_can.reset_index()

In [21]:
df.head(3)

Continent           Region             DevName  1980  1981  1982  \
Country                                                                        
Afghanistan      Asia    Southern Asia  Developing regions    16    39    39   
Albania        Europe  Southern Europe   Developed regions     1     0     0   
Algeria        Africa  Northern Africa  Developing regions    80    67    71   

             1983  1984  1985  1986  ...  2005  2006  2007  2008  2009  2010  \
Country                              ...                                       
Afghanistan    47    71   340   496  ...  3436  3009  2652  2111  1746  1758   
Albania         0     0     0     1  ...  1223   856   702   560   716   561   
Algeria        69    63    44    69  ...  3626  4807  3623  4005  5393  4752   

             2011  2012  2013  Total  
Country                               
Afghanistan  2203  2635  2004  58639  
Albania       539   620   603  15699  
Algeria      4325  3774  4331  69439  

[3 rows x 38 columns]

Optional: to remove the name of the index

df_can.index.name = None

Let's view the number of immigrants from Japan (row 87) for the following scenarios:
    1. The full row data (all columns)
    2. For year 2013
    3. For years 1980 to 1985


In [22]:
# 1. the full row data (all columns)        
df.loc['Japan'].head(10)

Continent                 Asia
Region            Eastern Asia
DevName      Developed regions
1980                       701
1981                       756
1982                       598
1983                       309
1984                       246
1985                       198
1986                       248
Name: Japan, dtype: object

Alternate methods

df_can.iloc[87] AND df_can[df_can.index == 'Japan']

In [23]:
# 2. for year 2013
df.loc['Japan', '2013']

982

In [24]:
# 3. for years 1980 to 1985
df.loc['Japan', ['1980', '1981', '1982', '1983', '1984', '1985']]

1980    701
1981    756
1982    598
1983    309
1984    246
1985    198
Name: Japan, dtype: object

Now Let's view the number of immigrants from **Haiti** for the following scenarios: <br>1. The full row data (all columns) <br>2. For year 2000 <br>3. For years 1990 to 1995

In [25]:
df.loc['Haiti'].head(10)

Continent    Latin America and the Caribbean
Region                             Caribbean
DevName                   Developing regions
1980                                    1666
1981                                    3692
1982                                    3498
1983                                    2860
1984                                    1418
1985                                    1321
1986                                    1753
Name: Haiti, dtype: object

In [26]:
df.loc['Haiti', '2000']

1631

In [27]:
df.loc['Haiti', ['1990', '1991', '1992', '1993', '1994', '1995']]

1990    2379
1991    2829
1992    2399
1993    3655
1994    2100
1995    2014
Name: Haiti, dtype: object

Column names that are integers (such as the years) might introduce some confusion. For example, when we are referencing the year 2013, one might confuse that when the 2013th positional index. 

To avoid this ambuigity, let's convert the column names into strings: '1980' to '2013'.


In [28]:
df.columns = list(map(str, df.columns))

Now let's check type of column headers.

In [ ]:
for x in df.columns.values:
    print(type(x))

Since we converted the years to string, let's declare a variable that will allow us to easily call upon the full range of years:
<br> [Useful for plotting later on.]

In [ ]:
years = list(map(str, range(1980,2014)))
years

Let's create a list named 'year' using map function for years ranging from 1990 to 2013. <br>Then extract the data series from the dataframe df_can for Haiti using year list. 

In [32]:
year = list(map(str, range(1990, 2014)))

In [33]:
haiti = df.loc['Haiti', year]
haiti

1990    2379
1991    2829
1992    2399
1993    3655
1994    2100
1995    2014
1996    1955
1997    1645
1998    1295
1999    1439
2000    1631
2001    2433
2002    2174
2003    1930
2004    1652
2005    1682
2006    1619
2007    1598
2008    2491
2009    2080
2010    4744
2011    6503
2012    5868
2013    4152
Name: Haiti, dtype: object

### Filtering based on a criteria <a id="4"></a>
To filter the dataframe based on a condition, we simply pass the condition as a boolean vector. 
<br> For example, Let's filter the dataframe to show the data on Asian countries (AreaName = Asia).


In [34]:
# 1. create the condition boolean series
df_asia = df['Continent'] == 'Asia'
print(df_asia)

Country
Afghanistan        True
Albania           False
Algeria           False
American Samoa    False
Andorra           False
                  ...  
Viet Nam           True
Western Sahara    False
Yemen              True
Zambia            False
Zimbabwe          False
Name: Continent, Length: 195, dtype: bool


In [35]:
# 2. pass this condition into the dataFrame
df[df_asia].head()

Continent         Region             DevName  1980  1981  1982  \
Country                                                                      
Afghanistan      Asia  Southern Asia  Developing regions    16    39    39   
Armenia          Asia   Western Asia  Developing regions     0     0     0   
Azerbaijan       Asia   Western Asia  Developing regions     0     0     0   
Bahrain          Asia   Western Asia  Developing regions     0     2     1   
Bangladesh       Asia  Southern Asia  Developing regions    83    84    86   

             1983  1984  1985  1986  ...  2005  2006  2007  2008  2009  2010  \
Country                              ...                                       
Afghanistan    47    71   340   496  ...  3436  3009  2652  2111  1746  1758   
Armenia         0     0     0     0  ...   224   218   198   205   267   252   
Azerbaijan      0     0     0     0  ...   359   236   203   125   165   209   
Bahrain         1     1     3     0  ...    12    12    22     9    35    28   
Bangladesh     81    98    92   486  ...  4171  4014  2897  2939  2104  4721   

             2011  2012  2013  Total  
Country                               
Afghanistan  2203  2635  2004  58639  
Armenia       236   258   207   3310  
Azerbaijan    138   161    57   2649  
Bahrain        21    39    32    475  
Bangladesh   2694  2640  3789  65568  

[5 rows x 38 columns]

We can pass multiple criteria in the same line.
<br> Let's filter for AreaNAme = Asia and RegName = Southern Asia

In [36]:
df[(df['Continent'] == 'Asia') & (df['Region'] == 'Southern Asia')]
# note: When using 'and' and 'or' operators, pandas requires we use '&' and '|' instead of 'and' and 'or'
# don't forget to enclose the two conditions in parentheses

Continent         Region             DevName  1980  \
Country                                                                         
Afghanistan                     Asia  Southern Asia  Developing regions    16   
Bangladesh                      Asia  Southern Asia  Developing regions    83   
Bhutan                          Asia  Southern Asia  Developing regions     0   
India                           Asia  Southern Asia  Developing regions  8880   
Iran (Islamic Republic of)      Asia  Southern Asia  Developing regions  1172   
Maldives                        Asia  Southern Asia  Developing regions     0   
Nepal                           Asia  Southern Asia  Developing regions     1   
Pakistan                        Asia  Southern Asia  Developing regions   978   
Sri Lanka                       Asia  Southern Asia  Developing regions   185   

                            1981  1982  1983  1984  1985  1986  ...   2005  \
Country                                                         ...          
Afghanistan                   39    39    47    71   340   496  ...   3436   
Bangladesh                    84    86    81    98    92   486  ...   4171   
Bhutan                         0     0     0     1     0     0  ...      5   
India                       8670  8147  7338  5704  4211  7150  ...  36210   
Iran (Islamic Republic of)  1429  1822  1592  1977  1648  1794  ...   5837   
Maldives                       0     0     1     0     0     0  ...      0   
Nepal                          1     6     1     2     4    13  ...    607   
Pakistan                     972  1201   900   668   514   691  ...  14314   
Sri Lanka                    371   290   197  1086   845  1838  ...   4930   

                             2006   2007   2008   2009   2010   2011   2012  \
Country                                                                       
Afghanistan                  3009   2652   2111   1746   1758   2203   2635   
Bangladesh                   4014   2897   2939   2104   4721   2694   2640   
Bhutan                         10      7     36    865   1464   1879   1075   
India                       33848  28742  28261  29456  34235  27509  30933   
Iran (Islamic Republic of)   7480   6974   6475   6580   7477   7479   7534   
Maldives                        0      2      1      7      4      3      1   
Nepal                         540    511    581    561   1392   1129   1185   
Pakistan                    13127  10124   8994   7217   6811   7468  11227   
Sri Lanka                    4714   4123   4756   4547   4422   3309   3338   

                             2013   Total  
Country                                    
Afghanistan                  2004   58639  
Bangladesh                   3789   65568  
Bhutan                        487    5876  
India                       33087  691904  
Iran (Islamic Republic of)  11291  175923  
Maldives                        1      30  
Nepal                        1308   10222  
Pakistan                    12603  241600  
Sri Lanka                    2394  148358  

[9 rows x 38 columns]

Now fetch the data where AreaName is 'Africa' and RegName is 'Southern Africa'. <br>Display the dataframe and find out how many instances are there?


In [37]:
df[(df['Continent'] =='Africa') & (df['Region'] =='Southern Asia')]

Empty DataFrame
Columns: [Continent, Region, DevName, 1980, 1981, 1982, 1983, 1984, 1985, 1986, 1987, 1988, 1989, 1990, 1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, Total]
Index: []

[0 rows x 38 columns]

### Sorting Values of a Dataframe or Series <a id="5"></a><br>
You can use the `sort_values()` function is used to sort a DataFrame or a Series based on one or more columns. <br>You to specify the column(s) by which you want to sort and the order (ascending or descending). Below is the syntax to use it:-<br><br>
```df.sort_values(col_name, axis=0, ascending=True, inplace=False, ignore_index=False)```<br><br>
col_nam - the column(s) to sort by. <br>
axis - axis along which to sort. 0 for sorting by rows (default) and 1 for sorting by columns.<br>
ascending - to sort in ascending order (True, default) or descending order (False).<br>
inplace - to perform the sorting operation in-place (True) or return a sorted copy (False, default).<br>
ignore_index - to reset the index after sorting (True) or keep the original index values (False, default).<br>


Let's sort out dataframe  on 'Total' column, in descending order to find out the top 5 countries that contributed the most to immigration to Canada. 


In [38]:
df.sort_values(by ='Total', ascending=False, axis =0, inplace=True)
top_5 = df.head()
top_5

Continent  \
Country                                                        
India                                                   Asia   
China                                                   Asia   
United Kingdom of Great Britain and Northern Ir...    Europe   
Philippines                                             Asia   
Pakistan                                                Asia   

                                                                Region  \
Country                                                                  
India                                                    Southern Asia   
China                                                     Eastern Asia   
United Kingdom of Great Britain and Northern Ir...     Northern Europe   
Philippines                                         South-Eastern Asia   
Pakistan                                                 Southern Asia   

                                                               DevName   1980  \
Country                                                                         
India                                               Developing regions   8880   
China                                               Developing regions   5123   
United Kingdom of Great Britain and Northern Ir...   Developed regions  22045   
Philippines                                         Developing regions   6051   
Pakistan                                            Developing regions    978   

                                                     1981   1982   1983  \
Country                                                                   
India                                                8670   8147   7338   
China                                                6682   3308   1863   
United Kingdom of Great Britain and Northern Ir...  24796  20620  10015   
Philippines                                          5921   5249   4562   
Pakistan                                              972   1201    900   

                                                     1984  1985  1986  ...  \
Country                                                                ...   
India                                                5704  4211  7150  ...   
China                                                1527  1816  1960  ...   
United Kingdom of Great Britain and Northern Ir...  10170  9564  9470  ...   
Philippines                                          3801  3150  4166  ...   
Pakistan                                              668   514   691  ...   

                                                     2005   2006   2007  \
Country                                                                   
India                                               36210  33848  28742   
China                                               42584  33518  27642   
United Kingdom of Great Britain and Northern Ir...   7258   7140   8216   
Philippines                                         18139  18400  19837   
Pakistan                                            14314  13127  10124   

                                                     2008   2009   2010  \
Country                                                                   
India                                               28261  29456  34235   
China                                               30037  29622  30391   
United Kingdom of Great Britain and Northern Ir...   8979   8876   8724   
Philippines                                         24887  28573  38617   
Pakistan                                             8994   7217   6811   

                                                     2011   2012   2013  \
Country                                                                   
India                                               27509  30933  33087   
China                                               28502  33024  34129   
United Kingdom of Great Britain and Northern Ir...   6204   6195   5827   
Philippines    

Now let's Find out top 3 countries that contributes the most to immigration to Canda in the year 2010. <br> Display the country names with the immigrant count in this year


In [39]:
df.sort_values(by ='2010', ascending =False, axis =0, inplace =True)
top_3_2010 = df['2010'].head(3)
top_3_2010

Country
Philippines    38617
India          34235
China          30391
Name: 2010, dtype: int64

Now let's Find out top 5 countries that contributes the most to immigration to Canda in the year 2000

In [44]:
df.sort_values(by ='2000', ascending=False, axis =0, inplace =True)
top_5_2000= df['2000'].head()
top_5_2000

Country
China                                                   35529
India                                                   28572
Pakistan                                                15400
Philippines                                             10763
United Kingdom of Great Britain and Northern Ireland     8840
Name: 2000, dtype: int64

Let's Find out the countries with *Least* immigration to Canada.

In [46]:
df.sort_values(by ='Total', ascending=False, axis =0, inplace =True)
bottom_5 = df.tail()
bottom_5

Continent           Region             DevName  1980  1981  \
Country                                                                       
San Marino          Europe  Southern Europe   Developed regions     1     0   
New Caledonia      Oceania        Melanesia  Developing regions     0     0   
Marshall Islands   Oceania       Micronesia  Developing regions     0     0   
Western Sahara      Africa  Northern Africa  Developing regions     0     0   
Palau              Oceania       Micronesia  Developing regions     0     0   

                  1982  1983  1984  1985  1986  ...  2005  2006  2007  2008  \
Country                                         ...                           
San Marino           0     0     0     0     1  ...     0     0     0     0   
New Caledonia        0     0     0     0     0  ...     0     0     0     1   
Marshall Islands     0     0     0     0     0  ...     0     0     2     0   
Western Sahara       0     0     0     0     0  ...     0     1     0     0   
Palau                0     0     0     0     0  ...     0     0     1     0   

                  2009  2010  2011  2012  2013  Total  
Country                                                
San Marino           0     1     0     0     0      5  
New Caledonia        0     0     0     0     2      5  
Marshall Islands     0     0     0     0     0      2  
Western Sahara       0     0     0     0     0      2  
Palau                0     0     0     0     0      1  

[5 rows x 38 columns]